### PR TITLE
add serviceexport controller with yacht

### PR DESCRIPTION
**Pull request by lmxia**
Signed-off-by: lmxia <xialingming@gmail.com>

#### What type of PR is this?
Feature Add
#### What this PR does / why we need it:
Service Export controller feature.
#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

Assume service export has been labeled with subscription namespace.

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -237,6 +250,15 @@ func (agent *Agent) run(ctx context.Context) {
 		}
 	}, time.Duration(0))
 
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		mcs.NewController(
```
Comment (dixudx) [inline pkg/agent/agent.go:254]: seems we could new this controller in `NewAgent`.

File: pkg/known/constant.go [hunk 2]
```diff
@@ -53,6 +53,7 @@ const (
 // These are internal finalizer values to Clusternet, must be qualified name.
 const (
 	AppFinalizer            string = "apps.clusternet.io/finalizer"
+	SeFinalizer             string = "se.clusternet.io/finalizer"
```
Comment (dixudx) [inline pkg/known/constant.go:56]: shall we use `AppFinalizer` instead?

or `apps.clusternet.io/serviceexports-finalizer`

File: pkg/agent/agent.go (hunk 1 above)
Comment (lmxia) [reply pkg/agent/agent.go:254]: Yes, that's semantics reasonable. 

But We need `parentDedicatedKubeConfig`, the field only availlable after agent register successfully. I think sockets.NewController is in the same case.

File: pkg/known/constant.go (hunk 2 above)
Comment (lmxia) [reply pkg/known/constant.go:56]: ok, AppFinalizer.

File: pkg/agent/agent.go (hunk 1 above)
Comment (lmxia) [reply pkg/agent/agent.go]: please take another look @dixudx

File: pkg/controllers/mcs/serviceexport_controller.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	seInformers.Informer().AddEventHandler(c.DefaultResourceEventHandlerFuncs())
+
+	return c
+}
+
+func (r *reconciler) Handle(obj interface{}) (requeueAfter *time.Duration, err error) {
+	ctx := context.Background()
+	key := obj.(string)
+	namespace, seName, err := cache.SplitMetaNamespaceKey(key)
+	utilruntime.Must(err)
```
Comment (dixudx) [inline pkg/controllers/mcs/serviceexport_controller.go:79]: This would lead to panic.

File: pkg/controllers/mcs/serviceexport_controller.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+	se, err := r.mcsClientset.MulticlusterV1alpha1().ServiceExports(namespace).Get(ctx, seName, metav1.GetOptions{})
```
Comment (dixudx) [inline pkg/controllers/mcs/serviceexport_controller.go:81]: Shall we use `mcsLister` to `Get` a resource?

File: pkg/agent/mcs.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	mcsclientset "sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+	mcsInformers "sigs.k8s.io/mcs-api/pkg/client/informers/externalversions"
+
+	"github.com/clusternet/clusternet/pkg/controllers/mcs"
+)
+
+type Mcs struct {
```
Comment (dixudx) [inline pkg/agent/mcs.go:29]: It seems we could combine this file with `serviceexport_controller.go`.

File: pkg/known/labels.go [hunk 6]
```diff
@@ -37,6 +37,8 @@ const (
 	ConfigSubscriptionUIDLabel       = "apps.clusternet.io/subs.uid"
 	ConfigSubscriptionNameLabel      = "apps.clusternet.io/subs.name"
 	ConfigSubscriptionNamespaceLabel = "apps.clusternet.io/subs.namespace"
+
+	LabelServiceName = "msc.clusternet.io/service-name"
```
[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/known/labels.go:41]: or `services.clusternet.io/multi-cluster-service-name`

File: pkg/utils/handler.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+*/
+package utils
+
+import (
+	"crypto/sha256"
+	"encoding/base32"
+	"strings"
+)
+
+func DerivedName(name string) string {
```
Comment (dixudx) [inline pkg/utils/handler.go:24]: We could add a test for this func.

File: pkg/controllers/mcs/serviceexport_controller.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+func (r *reconciler) constructEndpointSlice(slice *discoveryv1.EndpointSlice, se *v1alpha1.ServiceExport) *discoveryv1.EndpointSlice {
+	// mutate slice fields before upload to parent cluster.
+	newSlice := &discoveryv1.EndpointSlice{}
+	newSlice.AddressType = slice.AddressType
+	newSlice.Endpoints = slice.Endpoints
+	newSlice.Ports = slice.Ports
+	newSlice.Labels = make(map[string]string)
+
+	newSlice.GetLabels()[known.LabelServiceName] = se.Name
+	newSlice.GetLabels()[discoveryv1.LabelServiceName] = utils.DerivedName(se.Name)
```
Comment (dixudx) [inline pkg/controllers/mcs/serviceexport_controller.go:176]: We could just use `newSlice.Labels[xxx]` directly.

Comment (dixudx) [reply pkg/controllers/mcs/serviceexport_controller.go:176]: Shall we add a label for service namespace.

File: go.mod [hunk 9]
```diff
@@ -4,6 +4,7 @@ go 1.17
 
 require (
 	github.com/davecgh/go-spew v1.1.1
+	github.com/dixudx/yacht v0.2.2-0.20220726105322-ea7e50f0c678
```
[the remark this answers is not in the document]
Comment (lmxia) [reply go.mod:7]: ok

File: pkg/known/labels.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lmxia) [reply pkg/known/labels.go:41]: ok, done

File: pkg/agent/mcs.go (hunk 5 above)
Comment (lmxia) [reply pkg/agent/mcs.go:29]: Later, we will include `endpointslice_controller`, so Mcs may be a wrapper to these two controllers, what do you think about that.

File: pkg/controllers/mcs/serviceexport_controller.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+			d := time.Second
+			return &d, err
+		}
+		klog.Infof("service export %s has been recycled successfully", se.Name)
+		return nil, nil
+	}
+
+	var endpointSliceList *discoveryv1.EndpointSliceList
+	if endpointSliceList, err = r.localk8sClient.DiscoveryV1().EndpointSlices(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{discoveryv1.LabelServiceName: se.Name}).String(),
```
[the remark this answers is not in the document]
Comment (lmxia) [reply pkg/controllers/mcs/serviceexport_controller.go:124]: here it's local endpointslice, L105 and L176 are parent endpointslices.

File: pkg/controllers/mcs/serviceexport_controller.go (hunk 3 above)
Comment (lmxia) [reply pkg/controllers/mcs/serviceexport_controller.go:79]: done.

File: pkg/controllers/mcs/serviceexport_controller.go (hunk 4 above)
Comment (lmxia) [reply pkg/controllers/mcs/serviceexport_controller.go:81]: ok, done

File: pkg/controllers/mcs/serviceexport_controller.go [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+		if err != nil {
+			d := time.Second
+			return &d, err
+		}
+		klog.Infof("service export %s has been recycled successfully", se.Name)
+		return nil, nil
+	}
+
+	var endpointSliceList *discoveryv1.EndpointSliceList
+	if endpointSliceList, err = r.localk8sClient.DiscoveryV1().EndpointSlices(namespace).List(ctx, metav1.ListOptions{
```
[the remark this answers is not in the document]
Comment (lmxia) [reply pkg/controllers/mcs/serviceexport_controller.go:123]: ok, done

File: pkg/utils/handler.go (hunk 7 above)
Comment (lmxia) [reply pkg/utils/handler.go:24]: ok

File: pkg/controllers/mcs/serviceexport_controller.go (hunk 8 above)
Comment (lmxia) [reply pkg/controllers/mcs/serviceexport_controller.go:176]: add subscription namspace: from label: `apps.clusternet.io/subs.namespace`

**Review by dixudx — APPROVED** [review]
Nice work. Thanks.

**Pull request merged**